# Bellabeat Data Analysis Project

## Background

Bellabeat is a high-tech manufactuer focusng of women health since 2013. Their well-designed products includes the Bellabeat app, wellness trackers Leaf, Time and Spring, as well as the Bellabeat membership, which forms a strong ecosystem to empower women with the knowledge abour their own health and habits. Now, Bellabeat is seeking more opportunities to grow by learning the current smart device market. 

### Business Objectives
The key task of this report is to analyze FitBit Fitness Tracker Data to gain insights into how consumers are using the their smart devices and give high-level recommendation for Bellabeat's marketing stretegy. 

### Key Stakeholders
* Urška Sršen: Bellabeat’s cofounder and Chief Creative Officer
* Sando Mur: Mathematician, Bellabeat’s cofounder and key member of the Bellabeat executive team
* Bellabeat marketing analytics team: A team of data analysts guiding Bellabeat's marketing strategy.


## Data Preparation

### Import packages and dataset

In [111]:
# import packages 
import numpy as np 
import pandas as pd 
import matplotlib as plt 
import datetime as dt 

In [112]:
# import data 
activity = pd.read_csv("/Users/yuefei/Documents/Code/Bellabeat-data-analysis-project/Data/Fitabase Data 4.12.16-5.12.16/dailyActivity_merged.csv")
calories = pd.read_csv("/Users/yuefei/Documents/Code/Bellabeat-data-analysis-project/Data/Fitabase Data 4.12.16-5.12.16/dailyCalories_merged.csv")
intensities = pd.read_csv("/Users/yuefei/Documents/Code/Bellabeat-data-analysis-project/Data/Fitabase Data 4.12.16-5.12.16/dailyIntensities_merged.csv")
steps = pd.read_csv("/Users/yuefei/Documents/Code/Bellabeat-data-analysis-project/Data/Fitabase Data 4.12.16-5.12.16/dailySteps_merged.csv")
sleep = pd.read_csv("/Users/yuefei/Documents/Code/Bellabeat-data-analysis-project/Data/Fitabase Data 4.12.16-5.12.16/sleepDay_merged.csv")
weight = pd.read_csv("/Users/yuefei/Documents/Code/Bellabeat-data-analysis-project/Data/Fitabase Data 4.12.16-5.12.16/weightLogInfo_merged.csv")

#### Daily Activity

In [113]:
# Have a look at the first 5 rows and the basic information
activity.head()

Id ActivityDate  TotalSteps  TotalDistance  TrackerDistance  \
0  1503960366    4/12/2016       13162           8.50             8.50   
1  1503960366    4/13/2016       10735           6.97             6.97   
2  1503960366    4/14/2016       10460           6.74             6.74   
3  1503960366    4/15/2016        9762           6.28             6.28   
4  1503960366    4/16/2016       12669           8.16             8.16   

   LoggedActivitiesDistance  VeryActiveDistance  ModeratelyActiveDistance  \
0                       0.0                1.88                      0.55   
1                       0.0                1.57                      0.69   
2                       0.0                2.44                      0.40   
3                       0.0                2.14                      1.26   
4                       0.0                2.71                      0.41   

   LightActiveDistance  SedentaryActiveDistance  VeryActiveMinutes  \
0                 6.06                      0.0                 25   
1                 4.71                      0.0                 21   
2                 3.91                      0.0                 30   
3                 2.83                      0.0                 29   
4                 5.04                      0.0                 36   

   FairlyActiveMinutes  LightlyActiveMinutes  SedentaryMinutes  Calories  
0                   13                   328               728      1985  
1                   19                   217               776      1797  
2                   11                   181              1218      1776  
3                   34                   209               726      1745  
4                   10                   221               773      1863

In [114]:
activity.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 940 entries, 0 to 939
Data columns (total 15 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Id                        940 non-null    int64  
 1   ActivityDate              940 non-null    object 
 2   TotalSteps                940 non-null    int64  
 3   TotalDistance             940 non-null    float64
 4   TrackerDistance           940 non-null    float64
 5   LoggedActivitiesDistance  940 non-null    float64
 6   VeryActiveDistance        940 non-null    float64
 7   ModeratelyActiveDistance  940 non-null    float64
 8   LightActiveDistance       940 non-null    float64
 9   SedentaryActiveDistance   940 non-null    float64
 10  VeryActiveMinutes         940 non-null    int64  
 11  FairlyActiveMinutes       940 non-null    int64  
 12  LightlyActiveMinutes      940 non-null    int64  
 13  SedentaryMinutes          940 non-null    int64  
 14  Calories  

Check if there is any missing data in daily activity. Turns out no missing data in the activity table. 

In [115]:
# checking null values
missing_values = activity.isnull().sum()
missing_values

Id                          0
ActivityDate                0
TotalSteps                  0
TotalDistance               0
TrackerDistance             0
LoggedActivitiesDistance    0
VeryActiveDistance          0
ModeratelyActiveDistance    0
LightActiveDistance         0
SedentaryActiveDistance     0
VeryActiveMinutes           0
FairlyActiveMinutes         0
LightlyActiveMinutes        0
SedentaryMinutes            0
Calories                    0
dtype: int64

Check the user number in this dataset and their id

In [116]:
# checking the sample size and their unique user number
unique_id = np.unique(activity.Id)
unique_user_number = np.count_nonzero(unique_id)
unique_id, unique_user_number

(array([1503960366, 1624580081, 1644430081, 1844505072, 1927972279,
        2022484408, 2026352035, 2320127002, 2347167796, 2873212765,
        3372868164, 3977333714, 4020332650, 4057192912, 4319703577,
        4388161847, 4445114986, 4558609924, 4702921684, 5553957443,
        5577150313, 6117666160, 6290855005, 6775888955, 6962181067,
        7007744171, 7086361926, 8053475328, 8253242879, 8378563200,
        8583815059, 8792009665, 8877689391]),
 33)

The activity data can be observed: 
* The activity table has 15 columns and 940 rows
* No missing data
* The sample size is 33, which is quite small. 
* The issue here is that date is marked as `object`. 

Meanwhile, I would like to check the other tables to see whether the data type issue is common. 

#### Calories

In [117]:
# checking calories table
calories.head()

Id ActivityDay  Calories
0  1503960366   4/12/2016      1985
1  1503960366   4/13/2016      1797
2  1503960366   4/14/2016      1776
3  1503960366   4/15/2016      1745
4  1503960366   4/16/2016      1863

In [118]:
calories.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 940 entries, 0 to 939
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Id           940 non-null    int64 
 1   ActivityDay  940 non-null    object
 2   Calories     940 non-null    int64 
dtypes: int64(2), object(1)
memory usage: 22.2+ KB


In [119]:
calories_missing_values = calories.isnull().sum()
calories_missing_values

Id             0
ActivityDay    0
Calories       0
dtype: int64

#### Intensities

In [120]:
# Intensities table
intensities.head()

Id ActivityDay  SedentaryMinutes  LightlyActiveMinutes  \
0  1503960366   4/12/2016               728                   328   
1  1503960366   4/13/2016               776                   217   
2  1503960366   4/14/2016              1218                   181   
3  1503960366   4/15/2016               726                   209   
4  1503960366   4/16/2016               773                   221   

   FairlyActiveMinutes  VeryActiveMinutes  SedentaryActiveDistance  \
0                   13                 25                      0.0   
1                   19                 21                      0.0   
2                   11                 30                      0.0   
3                   34                 29                      0.0   
4                   10                 36                      0.0   

   LightActiveDistance  ModeratelyActiveDistance  VeryActiveDistance  
0                 6.06                      0.55                1.88  
1                 4.71                      0.69                1.57  
2                 3.91                      0.40                2.44  
3                 2.83                      1.26                2.14  
4                 5.04                      0.41                2.71

In [121]:
intensities.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 940 entries, 0 to 939
Data columns (total 10 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Id                        940 non-null    int64  
 1   ActivityDay               940 non-null    object 
 2   SedentaryMinutes          940 non-null    int64  
 3   LightlyActiveMinutes      940 non-null    int64  
 4   FairlyActiveMinutes       940 non-null    int64  
 5   VeryActiveMinutes         940 non-null    int64  
 6   SedentaryActiveDistance   940 non-null    float64
 7   LightActiveDistance       940 non-null    float64
 8   ModeratelyActiveDistance  940 non-null    float64
 9   VeryActiveDistance        940 non-null    float64
dtypes: float64(4), int64(5), object(1)
memory usage: 73.6+ KB


In [122]:
intensities_missing_values = intensities.isnull().sum()
intensities_missing_values

Id                          0
ActivityDay                 0
SedentaryMinutes            0
LightlyActiveMinutes        0
FairlyActiveMinutes         0
VeryActiveMinutes           0
SedentaryActiveDistance     0
LightActiveDistance         0
ModeratelyActiveDistance    0
VeryActiveDistance          0
dtype: int64

#### Steps

In [123]:
# steps table
steps.head()

Id ActivityDay  StepTotal
0  1503960366   4/12/2016      13162
1  1503960366   4/13/2016      10735
2  1503960366   4/14/2016      10460
3  1503960366   4/15/2016       9762
4  1503960366   4/16/2016      12669

In [124]:
steps.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 940 entries, 0 to 939
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Id           940 non-null    int64 
 1   ActivityDay  940 non-null    object
 2   StepTotal    940 non-null    int64 
dtypes: int64(2), object(1)
memory usage: 22.2+ KB


In [125]:
steps_missing_values = steps.isnull().sum()
steps_missing_values

Id             0
ActivityDay    0
StepTotal      0
dtype: int64

#### Sleep

In [126]:
# sleep
sleep.head()

Id               SleepDay  TotalSleepRecords  TotalMinutesAsleep  \
0  1503960366  4/12/2016 12:00:00 AM                  1                 327   
1  1503960366  4/13/2016 12:00:00 AM                  2                 384   
2  1503960366  4/15/2016 12:00:00 AM                  1                 412   
3  1503960366  4/16/2016 12:00:00 AM                  2                 340   
4  1503960366  4/17/2016 12:00:00 AM                  1                 700   

   TotalTimeInBed  
0             346  
1             407  
2             442  
3             367  
4             712

In [127]:
sleep.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 413 entries, 0 to 412
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Id                  413 non-null    int64 
 1   SleepDay            413 non-null    object
 2   TotalSleepRecords   413 non-null    int64 
 3   TotalMinutesAsleep  413 non-null    int64 
 4   TotalTimeInBed      413 non-null    int64 
dtypes: int64(4), object(1)
memory usage: 16.3+ KB


In [128]:
sleep_missing_values = sleep.isnull().sum()
sleep_missing_values

Id                    0
SleepDay              0
TotalSleepRecords     0
TotalMinutesAsleep    0
TotalTimeInBed        0
dtype: int64

#### Weight

In [129]:
# weight
weight.head()

Id                   Date    WeightKg  WeightPounds   Fat  \
0  1503960366   5/2/2016 11:59:59 PM   52.599998    115.963147  22.0   
1  1503960366   5/3/2016 11:59:59 PM   52.599998    115.963147   NaN   
2  1927972279   4/13/2016 1:08:52 AM  133.500000    294.317120   NaN   
3  2873212765  4/21/2016 11:59:59 PM   56.700001    125.002104   NaN   
4  2873212765  5/12/2016 11:59:59 PM   57.299999    126.324875   NaN   

         BMI  IsManualReport          LogId  
0  22.650000            True  1462233599000  
1  22.650000            True  1462319999000  
2  47.540001           False  1460509732000  
3  21.450001            True  1461283199000  
4  21.690001            True  1463097599000

In [130]:
weight.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 67 entries, 0 to 66
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Id              67 non-null     int64  
 1   Date            67 non-null     object 
 2   WeightKg        67 non-null     float64
 3   WeightPounds    67 non-null     float64
 4   Fat             2 non-null      float64
 5   BMI             67 non-null     float64
 6   IsManualReport  67 non-null     bool   
 7   LogId           67 non-null     int64  
dtypes: bool(1), float64(4), int64(2), object(1)
memory usage: 3.9+ KB


In [131]:
weight_missing_values = weight.isnull().sum()
weight_missing_values

Id                 0
Date               0
WeightKg           0
WeightPounds       0
Fat               65
BMI                0
IsManualReport     0
LogId              0
dtype: int64

After checking all the tables, it is confirmed that all the date are stored as object. 

The following date transformation will be performed: 
1. All the date is converted from `object` to `datetime`
2. The date is formatted to ``dd-mm-yyyy`` as commonly used in Australia
3. Date with daytime like `5/2/2016 11:59:59 PM` will be split into two columns

### Data Transformation

In [132]:
# convert the activity data time
activity["ActivityDate"] = pd.to_datetime(activity["ActivityDate"],format="%m/%d/%Y")

#check the updated activity data
activity.info()
activity["ActivityDate"].head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 940 entries, 0 to 939
Data columns (total 15 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   Id                        940 non-null    int64         
 1   ActivityDate              940 non-null    datetime64[ns]
 2   TotalSteps                940 non-null    int64         
 3   TotalDistance             940 non-null    float64       
 4   TrackerDistance           940 non-null    float64       
 5   LoggedActivitiesDistance  940 non-null    float64       
 6   VeryActiveDistance        940 non-null    float64       
 7   ModeratelyActiveDistance  940 non-null    float64       
 8   LightActiveDistance       940 non-null    float64       
 9   SedentaryActiveDistance   940 non-null    float64       
 10  VeryActiveMinutes         940 non-null    int64         
 11  FairlyActiveMinutes       940 non-null    int64         
 12  LightlyActiveMinutes  

0   2016-04-12
1   2016-04-13
2   2016-04-14
3   2016-04-15
4   2016-04-16
Name: ActivityDate, dtype: datetime64[ns]

In [133]:
# convert date of calories, intensities, steps, sleep, weight to datetime64 
# format only date
calories["ActivityDay"] = pd.to_datetime(calories["ActivityDay"],format="%m/%d/%Y")
intensities["ActivityDay"] = pd.to_datetime(intensities["ActivityDay"],format="%m/%d/%Y")
steps["ActivityDay"] = pd.to_datetime(steps["ActivityDay"],format="%m/%d/%Y")

# format date, time and am/pm
sleep["SleepDay"] = pd.to_datetime(sleep["SleepDay"],format="%m/%d/%Y %H:%M:%S %p")
weight["Date"] = pd.to_datetime(weight["Date"],format="%m/%d/%Y %I:%M:%S %p")

#check the time format in sleep and weight
weight["Date"].head()


0   2016-05-02 23:59:59
1   2016-05-03 23:59:59
2   2016-04-13 01:08:52
3   2016-04-21 23:59:59
4   2016-05-12 23:59:59
Name: Date, dtype: datetime64[ns]

In [134]:
# format date, time and am/pm
sleep["SleepDay"] = pd.to_datetime(sleep["SleepDay"],format="%m/%d/%Y %H:%M:%S %p")
weight["Date"] = pd.to_datetime(weight["Date"],format="%m/%d/%Y %I:%M:%S %p")

#check the time format in sleep and weight
sleep["SleepDay"].head()

0   2016-04-12 12:00:00
1   2016-04-13 12:00:00
2   2016-04-15 12:00:00
3   2016-04-16 12:00:00
4   2016-04-17 12:00:00
Name: SleepDay, dtype: datetime64[ns]

In [136]:
# check if 24hr format is working. 
weight["Date"].head()

0   2016-05-02 23:59:59
1   2016-05-03 23:59:59
2   2016-04-13 01:08:52
3   2016-04-21 23:59:59
4   2016-05-12 23:59:59
Name: Date, dtype: datetime64[ns]